File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        int m = matrix[0].length;

        int start = 0;
        int end = n*m-1;

        while(start <= end) {
            int mid = start + (end - start) /  2;
            int midVal = matrix[mid / m][mid % m];  

            if(midVal == target) {
                return true;
            }

            else if(target < midVal) {
                end = mid - 1;
            }
            else {
                start = mid + 1;
            }
        }
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        m = len(matrix[0])
        l = 0
        r = n*m-1
        while(l<=r):
            mid = (l+r)//2
            row = mid//m
            col = mid%m
            midEle = matrix[row][col]

            if(target < midEle):
                r = mid-1
            elif(target > midEle):
                l = mid+1
            else:
                return True
        return False